drivers/flash/flash_sam.c

```diff
@@ -1,404 +1,492 @@
+{
+	struct sam_flash_data *sam_data = data;
+	const struct device *dev = sam_data->dev;
+	struct sam_flash_erase_data *erase_data = &sam_data->erase_data;
+
+	/* Validate we reached first page to erase */
+	if (info->start_offset < erase_data->section_start) {
+		/* Next page */
+		return true;
 	}
 
-	int rc2 = flash_sam_write_protection(dev, true);
-
-	if (!rc) {
-		rc = rc2;
+	/* Check if we've reached the end of pages to erase */
+	if (info->start_offset >= erase_data->section_end) {
+		/* Succeeded, stop iterating */
+		erase_data->succeeded = true;
+		return false;
 	}
 
-	flash_sam_sem_give(dev);
+	if (sam_flash_unlock_page(dev, info) < 0) {
+		/* Failed to unlock page, stop iterating */
+		return false;
+	}
 
-	/*
-	 * Invalidate the cache addresses corresponding to the erased blocks,
-	 * so that they really appear as erased.
-	 */
-#ifdef __DCACHE_PRESENT
-	SCB_InvalidateDCache_by_Addr((void *)(CONFIG_FLASH_BASE_ADDRESS + offset), len);
-#endif
+	if (sam_flash_erase_page(dev, info) < 0) {
+		/* Failed to erase page, stop iterating */
+		return false;
+	}
 
-	return rc;
+	/* Next page */
+	return true;
 }
 
-/* Enable or disable the write protection */
-static int flash_sam_write_protection(const struct device *dev, bool enable)
+static int sam_flash_erase(const struct device *dev, off_t offset, size_t size)
 {
-#if defined(EFC_6450)
-	const struct flash_sam_dev_cfg *config = dev->config;
-	Efc *const efc = config->regs;
-#endif
-	int rc = 0;
-
-	if (enable) {
-		rc = flash_sam_wait_ready(dev);
-		if (rc < 0) {
-			goto done;
-		}
-#if defined(EFC_6450)
-		efc->EEFC_WPMR = EEFC_WPMR_WPKEY_PASSWD | EEFC_WPMR_WPEN;
-	} else {
-		efc->EEFC_WPMR = EEFC_WPMR_WPKEY_PASSWD;
-#endif
+	struct sam_flash_data *sam_data = dev->data;
+	k_spinlock_key_t key;
+
+	if (size == 0) {
+		return 0;
 	}
 
-done:
-	return rc;
-}
+	if (!sam_flash_section_is_aligned_with_pages(dev, offset, size)) {
+		return -EINVAL;
+	}
 
-#if CONFIG_FLASH_PAGE_LAYOUT
-/*
- * The notion of pages is different in Zephyr and in the SAM documentation.
- * Here a page refers to the granularity at which the flash can be erased.
- */
-static const struct flash_pages_layout flash_sam_pages_layout = {
-	.pages_count = DT_REG_SIZE(SOC_NV_FLASH_NODE) / FLASH_ERASE_BLK_SZ,
-	.pages_size = DT_PROP(SOC_NV_FLASH_NODE, erase_block_size),
-};
+	LOG_DBG("Erasing sector from 0x%x to 0x%x", (size_t)offset, (size_t)(offset + size));
 
-void flash_sam_page_layout(const struct device *dev,
-			   const struct flash_pages_layout **layout,
-			   size_t *layout_size)
-{
-	*layout = &flash_sam_pages_layout;
-	*layout_size = 1;
+	key = k_spin_lock(&sam_data->lock);
+	sam_data->erase_data.section_start = offset;
+	sam_data->erase_data.section_end = offset + size;
+	sam_data->erase_data.succeeded = false;
+	flash_page_foreach(dev, sam_flash_erase_foreach_page, sam_data);
+	if (!sam_data->erase_data.succeeded) {
+		k_spin_unlock(&sam_data->lock, key);
```

Comment:
you could probably have the unlock above if and discard the one below.

Reply:
The unlock needs to be after since the `erase_data` context is shared, and must be protected until read.